### PR TITLE
Exclude unmapped fields during max clause limit checking for querying

### PR DESCRIPTION
**Pull request by zacharymorn**
Take into account of number of unmapped fields when calculating against limit. For details, please see #49002

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java [hunk 1]
```diff
@@ -143,6 +152,7 @@ private QueryParserHelper() {}
             if (fieldType == null) {
                 // Note that we don't ignore unmapped fields.
                 fields.put(fieldName, weight);
+                unmappedFields.add(fieldName);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: We can ignore unmapped fields here and change the comment ? That's the scope of #49002, we want to ignore unmapped fields so we don't need to forward them to the Lucene parsing.

File: server/src/test/resources/config/elasticsearch.yml [hunk 2]
```diff
@@ -1,3 +1,2 @@
 
 yaml.config.exists: true
-
```
Comment (jimczi) [inline server/src/test/resources/config/elasticsearch.yml]: can you revert this change

File: test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java [hunk 3]
```diff
@@ -212,21 +212,14 @@ public static void afterClass() throws Exception {
```
Comment (jimczi) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Changes on this file are not needed. Can you revert ?

File: server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java [hunk 4]
```diff
@@ -1290,6 +1299,61 @@ public void testQuoteAnalyzer() throws Exception {
...
+        new QueryStringQueryBuilder("\"bar\"")
+                .field("*")
+                .field("unmappedField1")
+                .field("unmappedField2")
+                .field("unmappedField3")
+                .field("unmappedField4")
+                .toQuery(shardContext);
+    }
+
+    public void testToQueryExceedingMaxCaluseCountWithMappedFeilds() throws IOException {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java]: nit: 
```suggestion
    public void testToQueryExceedingMaxClauseCountWithMappedFields() throws IOException {
```

File: server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java [hunk 5]
```diff
@@ -1290,6 +1299,61 @@ public void testQuoteAnalyzer() throws Exception {
...
+                .toQuery(shardContext);
+            fail();
+        }
+        catch (IllegalArgumentException e) {
+            assertEquals("field expansion matches too many fields, limit: 13, got: 17", e.getMessage());
+        }
+
+    }
+
+    public void testToQueryExceedingMaxCaluseCountWithMappedFeildsWithoutWeightAndBoost() throws IOException {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java]: lot of words ;), 
```suggestion
    public void testToQueryMaxClauseNoWeight() throws IOException {
```
?

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks for your contribution @zacharymorn . I left some comments that need to be addressed. The pr could also be simplified to ignore unmapped fields completely rather than keeping the count in a separate `Set` (see review comment).

File: server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java (hunk 1 above)
Comment (zacharymorn) [reply server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java]: Got it. I thought previously they may still be needed down the pipe. I've updated the PR per your suggestion.

File: server/src/test/resources/config/elasticsearch.yml (hunk 2 above)
Comment (zacharymorn) [reply server/src/test/resources/config/elasticsearch.yml]: Ah missed this, removed.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java (hunk 3 above)
Comment (zacharymorn) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Sorry I forgot to add a comment for this. The changes here are because in two of the new test cases I added, the state of `shardContext.mapperService` was changed to force the condition, which affected other test cases as well. By re-creating `serviceHolder` for every test case, side effects from test case does not affect another.

File: server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java (hunk 4 above)
Comment (zacharymorn) [reply server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java]: Done.

File: server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java (hunk 5 above)
Comment (zacharymorn) [reply server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java]: Done.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java (hunk 3 above)
Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Why is it not possible to add your extra fields in `QueryStringQueryBuilderTests#initializeAdditionalMappings` ? That's where we expect tests to modify the shared mapping. Can you show the test failures if that doesn't work ?

File: server/src/test/resources/config/elasticsearch.yml (hunk 2 above)
Comment (jimczi) [reply server/src/test/resources/config/elasticsearch.yml]: It seems that it's still there ? ;)

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @zacharymorn , I left more comments

File: test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java (hunk 3 above)
Comment (zacharymorn) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: I did look at that method before. I think the problem of adding additional fields in `QueryStringQueryBuilderTests#initializeAdditionalMappings` was that it is adding additional mappings for ALL test cases in that class (the additional xContent mapping is hard coded in the method, which is an override one as well so I can't change its interface to take in additional parameter easily as well), so that may cause other test cases to fail in that test class as well.

File: server/src/test/resources/config/elasticsearch.yml (hunk 2 above)
Comment (zacharymorn) [reply server/src/test/resources/config/elasticsearch.yml]: Hmm that's strange, might be my IDE setup problem. Anyway I checked out and committed the previous version of this file before my changes, and I don't see this change in `Files Changed` tab anymore. so it should be good now.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java (hunk 3 above)
Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Sorry for the delay. I tried to modify `QueryStringQueryBuilderTests#initializeAdditionalMappings` like the following:
````
@Override
    protected void initializeAdditionalMappings(MapperService mapperService) throws IOException {
        XContentBuilder mapping = jsonBuilder().startObject().startObject("_doc").startObject("properties")
            .startObject("prefix_field")
                .field("type", "text")
                    .startObject("index_prefixes").endObject()
                .endObject()
            .startObject("additionalMappedField1")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField2")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField3")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField4")
                .field("type", "text")
            .endObject()
            .endObject().endObject().endObject();

        mapperService.merge("_doc",
            new CompressedXContent(Strings.toString(mapping)), MapperService.MergeReason.MAPPING_UPDATE);
    }
`````
and all tests worked without modification. Could you  try so that we can remove this  unneeded change ?

Comment (zacharymorn) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Sorry for the delay. I tried to implement what you suggested and pushed the changes to a different branch / commit, available to review here:  
Branch: https://github.com/zacharymorn/elasticsearch/commits/issue-49002-experiment
Commit: https://github.com/zacharymorn/elasticsearch/commit/e39bcdf27b91dd55c3eb77fbdc7bf04fb1040f8b

I ran the updated test cases in QueryStringQueryBuilderTests and many test cases did fail now because of the uniform initialization of additional fields for each test case:

Screen Shot 2020-01-08 at 10 10 54 PM

Am I missing anything here ?

Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Yes sorry I checked without the modified `indices.query.bool.max_clause_count`. It seems tricky to test this behavior here so maybe you can modify or add a new test similar to https://github.com/elastic/elasticsearch/blob/a59b065091a0bec7b9e66ed3ae0e878562947cf8/server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java#L266 ?

Comment (zacharymorn) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: Didn’t realize there’s an integration test for this as well…I just added a new test there. 

I also removed this line of Index setting https://github.com/elastic/elasticsearch/blob/a59b065091a0bec7b9e66ed3ae0e878562947cf8/server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java#L279-L280 in test `QueryStringIT.testLimitOnExpandedFields` as it doesn’t seems to make sense in the context (why  is it setting INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING  to CLUSTER_MAX_CLAUSE_COUNT + 100?), and have no impact to test result. But please let me know if that’s indeed needed and I can put it back.

Having added the additional test though, I feel this test and the ones I’ve already put in and passing are different levels of tests (integration & unit tests) and thus it doesn’t hurt for them to coexist.  Are you suggesting to have integration test only for these changes?

Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: > Are you suggesting to have integration test only for these changes?

Yes, mainly because it seems challenging to have a simple unit test for this and we already have integration tests.

> In test QueryStringIT.testLimitOnExpandedFields as it doesn’t seems to make sense in the context (why is it setting INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING to CLUSTER_MAX_CLAUSE_COUNT + 100?), and have no impact to test result. But please let me know if that’s indeed needed and I can put it back.

I don't think it's needed either but that's unrelated to this pr so not sure if we should remove it now ?

Comment (zacharymorn) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: I just pushed another commit that reverted the unneeded change to the integration test, and the unit tests and their related changes.

However, I guess I’m still a bit confused for why it’s tricky to keep those unit tests there, when they are already passing and verifying the right logic. As can be seen in the new (reverting) commit, the changes I made to allow unit tests  there are the following:
1. The new unit tests themselves
2. Local, test class level change to force the condition for the added unit tests (setting `indices.query.bool.max_clause_count`)
3. Super test class level change to make sure that each tests are independent to each other so that #2 doesn’t break other test cases (the changes in `AbstractBuilderTestCase.beforeTest`)

These changes combined would allow the newly added test cases to pass, and not break any existing test cases.

I feel I’m still having some gap in understanding the reasoning here, but I’ll leave it to you. Thanks for your patience and time for reviewing my changes!

Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java]: > I feel I’m still having some gap in understanding the reasoning here, but I’ll leave it to you

My reasoning here is that the abstract class for these tests was designed to share the mapping within the tests so it would require a bigger refactoring to change this behavior. We also have a lot of unit tests here so even if they pass I am cautious of the additional work that would be needed if we switch to one configuration per test.

**Comment by jimczi** [comment]
@elasticmachine ok to test.

**Comment by jimczi** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java [hunk 6]
```diff
@@ -295,6 +295,35 @@ public void testLimitOnExpandedFields() throws Exception {
...
+        builder.startObject("type1");
+        builder.startObject("properties");
+        for (int i = 0; i < CLUSTER_MAX_CLAUSE_COUNT; i++) {
+            builder.startObject("field" + i).field("type", "text").endObject();
+        }
+        builder.endObject(); // properties
+        builder.endObject(); // type1
+        builder.endObject();
+
+        assertAcked(prepareCreate("ignoreunmappedfields").addMapping("type1", builder));
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: @zacharymorn I merged master in your pr and now there's a test compilation error. Can you take a look ?

Comment (zacharymorn) [reply server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: Looks like there’s change to `CreateIndexRequestBuilder.addMapping` method . I just updated the test and pushed up the commit. It should work for CI now.

Comment (zacharymorn) [reply server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: I merged master in and looks like there’s change in  `CreateIndexRequestBuilder.addMapping` . Just updated the test and pushed up the commits.  The CI is still failing though it was broken for an seemingly unrelated integration test `org.elasticsearch.snapshots.DedicatedClusterSnapshotRestoreIT.testMasterShutdownDuringSnapshot`. Is this something I should be looking into as well ?

**Comment by jimczi** [comment]
@elasticmachine test elasticsearch-ci/1

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Comment by jimczi** [comment]
@elasticmachine update branch

**Review by jimczi — APPROVED** [review]

**Comment by jimczi** [comment]
Thanks @zacharymorn
